### PR TITLE
funcx ws draft

### PR DESCRIPTION
**Pull request by Loonride**

### CODE REVIEW

File: funcx_web_service/routes/funcx.py [hunk 1]
```diff
@@ -212,21 +212,33 @@ def submit(user: User):
         # this should raise a 500 because it prevented any tasks from launching
         return create_error_response(RequestKeyError(e), jsonify_response=True)
 
+    batch_id = str(uuid.uuid4())
```
Comment (yadudoc) [inline funcx_web_service/routes/funcx.py:215]: We could try to parse the `post_req` for a `taskgroup_id` field and if it's missing mint an id. I'm not sure if we want a situation where we do not create a batch at all.

File: funcx_web_service/routes/funcx.py [hunk 2]
```diff
@@ -212,21 +212,33 @@ def submit(user: User):
...
+            success_count += 1
+        else:
+            # the response code is a 207 if some tasks failed to submit
             final_http_status = 207
 
         results['results'].append(res)
+
+    if success_count > 0:
+        rc = g_redis_client()
+        Batch(rc, batch_id, user_id, success_count)
```
Comment (yadudoc) [inline funcx_web_service/routes/funcx.py:240]: We probably will need to tweak this to make the Batch object updatable.

**Review by yadudoc — COMMENTED** [review]
I'm adding some comments from the perspective of allowing the current batch_id notion to be easier to extend.

File: funcx_web_service/models/tasks.py [hunk 3]
```diff
@@ -170,3 +176,60 @@ def from_id(cls, rc: StrictRedis, task_id: str):
...
+
+
+@auto_name_fields
+class TaskGroup:
+    """
+    ORM-esque class to wrap access to properties of batches for better style and encapsulation
+    """
+    user_id = RedisField(serializer=lambda x: str(x), deserializer=lambda x: int(x))
+
+    TASK_GROUP_TTL = timedelta(weeks=1)
```
[the remark this answers is not in the document]
Comment (yadudoc) [reply funcx_web_service/models/tasks.py:188]: We should probably add a comment here, and come back to this.

**Review by yadudoc — APPROVED** [review]
Let's convert this to a PR and get this merged. This should not break the client.
We probably should update the version number to indicate that something has changed on the funcx-web-service.

File: funcx_web_service/models/tasks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yadudoc) [reply funcx_web_service/models/tasks.py:188]: After discussion with Ryan and ZZ, we should set this to 2 weeks, and come back if someone complains.

File: funcx_web_service/routes/funcx.py [hunk 4]
```diff
@@ -212,21 +221,41 @@ def submit(user: User):
         # this should raise a 500 because it prevented any tasks from launching
         return create_error_response(RequestKeyError(e), jsonify_response=True)
 
+    rc = g_redis_client()
+    task_group = None
+    if task_group_id and TaskGroup.exists(rc, task_group_id):
+        app.logger.debug(f'Task Group {task_group_id} submitted to by user {user_id} already exists, checking if user is authorized')
+        task_group = TaskGroup.from_id(rc, task_group_id)
```
Comment (yadudoc) [inline funcx_web_service/routes/funcx.py:228]: We should add a comment here that in the future we could get this from a cached function to speed this lookup.

**Review by ryanchard — APPROVED** [review]
Looks good!

**Pull request merged**